### DISCUSSION TITLE
How to disable background formatting…?

### DISCUSSION THREAD

**Discussion by hedefalk** [scalameta/metals#6976] [General]
I'm using metals from zed and I'm currently working in a monorepo on entirely unrelated code. metals seem to trigger formatting in the background of files I've never even opened a buffer for. How do I disable this?

Sonnet suggested it was the presentation compiler and that "enableScalafmtOnBuild" should fix it but seems to be hallucination :)

**Comment by tgodzik** [comment]
Metals doesn't format files unless explicitly asked by the editor to do so. The request to do it actually works on a single file, we never format anything on our own.

**Comment by hedefalk** [comment]
>we never format anything on our own.

Thanks, good to know and that figures. Been away from this for too long to have a good idea how to debug this right now. Any help here is highly appreciated!

Fact remains: as soon as I open this folder in zed, some scala files get formatted and I get git diffs. I'm using sbt instead of bloop on because I never got bloop to work well for my projects. Now I'm on zed and my current settings look like:

 ```
 "lsp": {
    "metals": {
      "settings": {
        "defaultBspToBuildTool": true,
        "inlayHints": {
          "inferredTypes": {
            "enable": false
          }
        },
        "binary": {
          "arguments": ["-Dmetals.http=on","-Dmetals.enableScalafmtOnBuild=false"]
        },
        "initialization_options": {
          "isHttpEnabled": true
        }
      }
    },
```

I just found that the root build.sbt _does_ have:

`scalafmtOnCompile                              := true,`

so I shouldn't be surprised, but there seem to be some mishap about the .scalafmt settings then, because if I run sbt compile from my terminal on the monorepo, nothing happens — and it shouldn't. Lots of people work on this code with this same scalafmt file.

But it happens straight away as soon as I open zed on the folder, even without any open buffers.


I see now in the metals lsp logs that:

>no build target found for /Users/viktor/[theproject]/build.sbt. Using presentation compiler with project's scala-library version: 3.3.3

We're using Scala 2.13.14


Then
```
...
Connected to Build server: sbt v1.10.4
…

2024.11.26 13:40:00 INFO  time: Imported build in 2.06s
2024.11.26 13:40:00 INFO  running doctor check
2024.11.26 13:40:00 INFO  java targets:
2024.11.26 13:40:00 WARN  Unable to run doctor. Make sure `isHttpEnabled` is set to `true`.
```
I thought I did that with 

"initialization_options": {
          "isHttpEnabled": true


Then:
```
…
2024.11.26 13:40:07 ERROR SemanticDB not found: build.sbt
scala.meta.internal.metals.ScalafixProvider$ScalafixRunException: SemanticDB not found: build.sbt

…

2024.11.26 13:42:07 INFO  scalafmt: Formatting 1 Scala sources (/Users/viktor/dev/[theproject]/[module])...

```


I think I had the same problem in vscode.

**Reply by tgodzik** [reply]
Wouldn't that be caused by `scalafmtOnCompile := true,` then? Maybe the scope is different, could it be that running from command line it only formats the changed files?

Does it format things if you do sbt compile?

**Reply by hedefalk** [reply]
Yes, I mean

scalafmtOnCompile := true

will cause it to format on compile (doh) but I was _sure_ it didn't change it up unless metals. But I need to double check…

And aah,

>no build target found for /Users/viktor/[theproject]/build.sbt. Using presentation compiler with project's scala-library version: 3.3.3

is a red herring, meaning only the meta build of the build.sbt file I guess.

**Reply by hedefalk** [reply]
Sorry, it seems now that `sbt compile` on the right _submodules_ will actually cause these reformats to happen. Nothing about metals then! Thanks.
